tests/io/dask/dataframe/test_repartition.py

```diff
@@ -0,0 +1,114 @@
+    dataset_factory_function_scope, df, target_nrows_per_partition, frac
+):
+    """
+    Because we carefully choose the number of original partitions and the value of
+    `target_nrows_per_partition`, we get exact matches for `target_nrows_per_partition`.
+    This is not expected in a real-world scenario where the conditions are very unlikely.
+    """
+    dfs = [df]
+    store_dataframes_as_dataset__iter(
+        dfs,
+        store=dataset_factory_function_scope.store_factory,
+        dataset_uuid=dataset_factory_function_scope.dataset_uuid,
+        partition_on=["part"],
+    )
+
+    repartition_ratio = calculate_repartition_ratio(
+        dataset_factory=dataset_factory_function_scope,
+        table_name=SINGLE_TABLE,
+        dataset_metadata_frac=frac,
+        target_nrows_per_partition=target_nrows_per_partition,
+    ).compute()
+
+    ddf = read_dataset_as_ddf(factory=dataset_factory_function_scope)
+    # call optimize here and find the ddf length
+    ddf_before_repartition = dask.optimize(
+        ddf, optimizations=[dask.dataframe.optimize]
+    )[0]
+
+    ddf = repartition_ddf(ddf, repartition_ratio=repartition_ratio)
+    # length of ddf after repartition should be less than before, to test the block fusing
+    # but if repartition_ratio==1, then the length should remain same.
+    assert len(ddf.dask.keys()) <= len(ddf_before_repartition.dask.keys())
+
+    assert (ddf.map_partitions(len).compute() == target_nrows_per_partition).all()
+
+
+@pytest.mark.parametrize("frac", (0.3, 0.1))
+@pytest.mark.parametrize(
+    "df,target_nrows_per_partition",
+    (
+        # repartition by decreasing partition size, increasing number of partitions
+        (pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25}), 97),
+        (pd.DataFrame({"A": range(100), "part": ["A", "B", "C", "D"] * 25}), 31),
+        # repartition by increasing partition size, decreasing number of partitions
+        (pd.DataFrame({"A": range(2 ** 6), "part": list("0123456789abcdef") * 4}), 11),
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 7),
+        (pd.DataFrame({"A": range(10), "part": range(10)}), 1),
+    ),
+)
+def test_calculate_repartition_ratio_approximate_match(
+    dataset_factory_function_scope, df, target_nrows_per_partition, frac
+):
+    dfs = [df]
+    store_dataframes_as_dataset__iter(
+        dfs,
+        store=dataset_factory_function_scope.store_factory,
+        dataset_uuid=dataset_factory_function_scope.dataset_uuid,
+        partition_on=["part"],
+    )
+
+    repartition_ratio = calculate_repartition_ratio(
+        dataset_factory=dataset_factory_function_scope,
+        table_name=SINGLE_TABLE,
+        dataset_metadata_frac=frac,
+        target_nrows_per_partition=target_nrows_per_partition,
+    ).compute()
+
+    ddf = read_dataset_as_ddf(factory=dataset_factory_function_scope)
+    # call optimize here and find the ddf length
+    ddf_before_repartition = dask.optimize(
```

Comment:
FYI: This optimize call is not sufficient to merge the repartition blocks. This is why we have the additional configuration in the `reparititon_ddf` function

Reply:
Have reverted the  changes in old testcases . Have added new testcase to test repartition ratio.